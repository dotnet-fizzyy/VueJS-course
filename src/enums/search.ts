export enum SearchByOptionNames {
    Title = 'title',
    Genre = 'genres',
}

export enum SortByOptionsNames {
    ReleaseDate = 'release_date',
    Rating = 'rating',
}

export enum SearchQueryParams {
    SearchTerm = 'term',
    SearchByOption = 'searchBy',
    SortByOption = 'sortBy',
}
