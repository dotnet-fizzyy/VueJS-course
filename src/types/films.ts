export interface Film {
    id: string;
    name: string;
    releaseYear: number;
    posterUrl?: string;
    posterAlt?: string;
    shortDescription: string;
}

export interface FilmShortDescription extends Film {}

export interface FilmFullDescription extends Film {
    rating: number;
    runtime: number;
    fullDescription: string;
}
