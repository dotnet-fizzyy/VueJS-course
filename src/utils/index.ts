import { FilmGetterKeys } from '@/enums/vuex';

export const castToString = (value: unknown): string => value?.toString() ?? '';

export const castNumberToStringWithPrecision = (value: number, digits: number = 1): string =>
    value.toLocaleString('en-US', {
        useGrouping: false,
        minimumFractionDigits: digits,
        maximumFractionDigits: digits,
    });

export const castNumberToStringWithRound = (value: number, direction: 'up' | 'down'): string => {
    switch (direction) {
        case 'up':
            return Math.ceil(value).toString();
        case 'down':
            return Math.floor(value).toString();
        default:
            return value.toString();
    }
};

export const getGetterKey = (module: string, key: FilmGetterKeys): string => `${module}/${key}`;
