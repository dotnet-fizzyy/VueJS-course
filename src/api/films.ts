import Vue from 'vue';
import VueResource from 'vue-resource';
import { CollectionResponse, Film, FilmResponse } from '@/types/films';
import { mapFilmApiResponseToFilm } from '@/utils/films';

export default class FilmsApi {
    private static readonly Headers: Record<string, string> = {
        Accept: 'application/json',
        'Content-Type': 'application/json',
    };

    public static async getFilms(): Promise<CollectionResponse<Film>> {
        const response: VueResource.HttpResponse = await Vue.http.get(`${FilmsApi.getBaseUrl()}/movies`, {
            headers: FilmsApi.Headers,
        });
        const responseBody: CollectionResponse<FilmResponse> = await response.json();

        return {
            total: responseBody.total,
            limit: responseBody.limit,
            offset: responseBody.offset,
            data: responseBody.data.map(mapFilmApiResponseToFilm),
        } as CollectionResponse<Film>;
    }

    public static async getFilmById(id: string): Promise<Film> {
        const response: VueResource.HttpResponse = await Vue.http.get(`${FilmsApi.getBaseUrl()}/movies/${id}`, {
            headers: FilmsApi.Headers,
        });
        const responseBody: FilmResponse = await response.json();

        return mapFilmApiResponseToFilm(responseBody) as Film;
    }

    private static getBaseUrl(): string {
        const baseUrl = Vue.http?.options?.root;

        if (!baseUrl) {
            throw new Error('Base URL for communicating with API is not defined');
        }

        return baseUrl;
    }
}
