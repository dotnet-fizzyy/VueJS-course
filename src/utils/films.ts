import { Film, FilmPreview, FilmResponse } from '@/types/films';
import { castToString, getYearFromStringDate } from '@/utils/index';

export const mapFilmToFilmPreview = (film: Film): FilmPreview => ({
    id: film.id,
    name: film.name,
    releaseYear: film.releaseYear,
    posterUrl: film.posterUrl,
    posterAlt: film.posterAlt,
    shortDescription: film.shortDescription,
});

export const mapFilmApiResponseToFilm = (filmRes: FilmResponse): Film => ({
    id: castToString(filmRes.id),
    name: filmRes.title,
    releaseYear: getYearFromStringDate(filmRes.release_date),
    posterUrl: filmRes.poster_path,
    posterAlt: filmRes.title.toLowerCase(),
    shortDescription: filmRes.label ?? 'Unspecified',
    fullDescription: filmRes.overview ?? 'Unspecified',
    rating: filmRes.vote_average ?? 0,
    runtime: filmRes.runtime ?? 0,
    genres: filmRes.genres,
});
