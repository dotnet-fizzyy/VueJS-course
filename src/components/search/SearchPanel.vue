<template>
    <div data-aqa-search-panel :class="$style.root">
        <span :class="$style['main-label']">Find your movie</span>

        <div :class="$style['search-container']">
            <div :class="$style['input-container']">
                <app-input
                    :value="searchTerm"
                    placeholder="Search"
                    :font-size="inputFontSize"
                    @change="onChangeSearchQuery"
                />
            </div>
            <div :class="$style['search-button-container']">
                <primary-button :font-size="inputFontSize" @click="onSearch">Search</primary-button>
            </div>
        </div>

        <div :class="$style['options-container']">
            <button-group
                label="Search By"
                :options="filterOptions"
                :selected-option="selectedSearchByOption"
                @select-option="onChangeSearchByOption"
            />
        </div>
    </div>
</template>

<script lang="ts">
import AppInput from '@/components/common/AppInput.vue';
import ButtonGroup from '@/components/common/ButtonGroup.vue';
import PrimaryButton from '@/components/common/PrimaryButton.vue';
import Vue from 'vue';
import { FilmGetterProps } from '@/vuex/modules/films/getters';
import { SearchByOptionNames } from '@/enums/search';
import { SearchByOptions } from '@/constants/search';
import {
    changeSearchActionPayload,
    changeSearchByActionPayload,
    searchFilmsActionPayload,
} from '@/vuex/modules/films/actions';
import { getFilmModuleType } from '@/vuex/store/utils';
import { mapGetters } from 'vuex';

export interface SearchPanelProps {
    onSearch: (searchValue: string, selectedFilterOption: string) => void;
}

export default Vue.extend({
    name: 'SearchPanel',
    components: { AppInput, PrimaryButton, ButtonGroup },
    data() {
        return {
            filterOptions: SearchByOptions,
            inputFontSize: 24,
        };
    },
    computed: {
        ...mapGetters({
            searchTerm: getFilmModuleType(FilmGetterProps.SearchTerm),
            selectedSearchByOption: getFilmModuleType(FilmGetterProps.SearchBy),
        }),
    },
    methods: {
        onSearch(): void {
            this.$store.dispatch(searchFilmsActionPayload());
        },
        onChangeSearchByOption(option: string): void {
            this.$store.dispatch(changeSearchByActionPayload(option as SearchByOptionNames));
        },
        onChangeSearchQuery(value: string): void {
            this.$store.dispatch(changeSearchActionPayload(value));
        },
    },
});
</script>

<style lang="scss" module scoped>
@import '../../styles/fonts';

.root {
    width: 100%;
    display: flex;
    flex-direction: column;
}

.main-label {
    @extend .font;

    text-transform: uppercase;
    font-size: 30px;
    color: $white;
    letter-spacing: 4px;
}

.search-container {
    margin-top: 50px;
    height: 60px;
    display: flex;
    flex-direction: row;
}

.input-container {
    flex-grow: 1;
}

.search-button-container {
    margin-left: 15px;
    flex-basis: 240px;
}

.options-container {
    margin-top: 15px;
    width: fit-content;
}
</style>
