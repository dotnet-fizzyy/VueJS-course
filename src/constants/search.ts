import { ButtonGroupOption } from '@/types/button';
import { SearchByOptionNames, SortByOptionsNames } from '@/enums/search';

export const SortByOptions: ButtonGroupOption[] = [
    { label: 'Release Date', name: SortByOptionsNames.ReleaseDate },
    { label: 'Rating', name: SortByOptionsNames.Rating },
];

export const SearchByOptions: ButtonGroupOption[] = [
    { label: 'Title', name: SearchByOptionNames.Title },
    { label: 'Genre', name: SearchByOptionNames.Genre },
];
