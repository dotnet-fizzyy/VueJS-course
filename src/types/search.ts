import { SearchQueryParams } from '@/enums/search';

export interface SearchParams {
    [SearchQueryParams.SearchTerm]?: string;
    [SearchQueryParams.SearchByOption]: string;
    [SearchQueryParams.SortByOption]: string;
}
