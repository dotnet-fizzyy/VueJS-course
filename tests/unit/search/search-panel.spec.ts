import SearchPanel from '@/components/search/SearchPanel.vue';
import Vuex from 'vuex';
import {
    FilmActionTypes,
    changeSearchActionPayload,
    changeSearchByActionPayload,
    getFilmsRequestActionPayload,
} from '@/vuex/modules/films/actions';
import { FilmGetterProps } from '@/vuex/modules/films/getters';
import { SearchByOptionNames } from '@/enums/search';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import { getFilmModuleType } from '@/vuex/store/utils';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('SearchPanel.vue Tests', () => {
    it('Should render', () => {
        //Arrange
        const actions = {
            [getFilmModuleType(FilmActionTypes.ChangeSortBy)]: jest.fn(),
        };
        const getters = {
            [getFilmModuleType(FilmGetterProps.SearchTerm)]: (): string => 'value',
            [getFilmModuleType(FilmGetterProps.SearchBy)]: (): SearchByOptionNames => SearchByOptionNames.Genre,
        };

        const store = new Vuex.Store({
            actions,
            getters,
        });

        store.dispatch = jest.fn();

        const wrapper = shallowMount(SearchPanel, {
            store,
            localVue,
        });

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should dispatch "getFilmsRequestActionPayload" action on clicking search button', () => {
        //Arrange
        const actions = {
            [getFilmModuleType(FilmActionTypes.ChangeSortBy)]: jest.fn(),
        };
        const getters = {
            [getFilmModuleType(FilmGetterProps.SearchTerm)]: (): string => 'value',
            [getFilmModuleType(FilmGetterProps.SearchBy)]: (): SearchByOptionNames => SearchByOptionNames.Genre,
        };

        const store = new Vuex.Store({
            actions,
            getters,
        });

        store.dispatch = jest.fn();

        const wrapper = shallowMount(SearchPanel, {
            store,
            localVue,
        });

        //Act & Assert
        const searchButtonWrapper = wrapper.findComponent({ name: 'primary-button' });

        expect(searchButtonWrapper.element).toBeTruthy();

        searchButtonWrapper.vm.$emit('click');

        expect(store.dispatch).toHaveBeenCalledWith(getFilmsRequestActionPayload());
    });

    it('Should dispatch "onChangeSearchQuery" action on changing search input', () => {
        //Arrange
        const actions = {
            [getFilmModuleType(FilmActionTypes.ChangeSortBy)]: jest.fn(),
        };
        const getters = {
            [getFilmModuleType(FilmGetterProps.SearchTerm)]: (): string => 'value',
            [getFilmModuleType(FilmGetterProps.SearchBy)]: (): SearchByOptionNames => SearchByOptionNames.Genre,
        };

        const store = new Vuex.Store({
            actions,
            getters,
        });

        store.dispatch = jest.fn();

        const wrapper = shallowMount(SearchPanel, {
            store,
            localVue,
        });

        const expectedSearchValue: string = 'new_search_value';

        //Act & Assert
        const searchAppInputWrapper = wrapper.findComponent({ name: 'app-input' });

        expect(searchAppInputWrapper.element).toBeTruthy();

        searchAppInputWrapper.vm.$emit('change', expectedSearchValue);

        expect(store.dispatch).toHaveBeenCalledWith(changeSearchActionPayload(expectedSearchValue));
    });

    it('Should dispatch "onChangeSearchByOption" action on changing search by option', () => {
        //Arrange
        const actions = {
            [getFilmModuleType(FilmActionTypes.ChangeSortBy)]: jest.fn(),
        };
        const getters = {
            [getFilmModuleType(FilmGetterProps.SearchTerm)]: (): string => 'value',
            [getFilmModuleType(FilmGetterProps.SearchBy)]: (): SearchByOptionNames => SearchByOptionNames.Genre,
        };

        const store = new Vuex.Store({
            actions,
            getters,
        });

        store.dispatch = jest.fn();

        const wrapper = shallowMount(SearchPanel, {
            store,
            localVue,
        });

        const expectedSearchByValue: SearchByOptionNames = SearchByOptionNames.Title;

        //Act & Assert
        const searchByOptionsWrapper = wrapper.findComponent({ name: 'button-group' });

        expect(searchByOptionsWrapper.element).toBeTruthy();

        searchByOptionsWrapper.vm.$emit('select-option', expectedSearchByValue);

        expect(store.dispatch).toHaveBeenCalledWith(changeSearchByActionPayload(expectedSearchByValue));
    });

    it('Should bind "searchTerm" getter to search input value', () => {
        //Arrange
        const searchTerm: string = 'SomeNewAwesomeValue';

        const getters = {
            [getFilmModuleType(FilmGetterProps.SearchTerm)]: (): string => searchTerm,
            [getFilmModuleType(FilmGetterProps.SearchBy)]: (): SearchByOptionNames => SearchByOptionNames.Genre,
        };

        const store = new Vuex.Store({
            getters,
        });

        const wrapper = shallowMount(SearchPanel, {
            store,
            localVue,
        });

        //Act & Assert
        const searchAppInputWrapper = wrapper.findComponent({ name: 'app-input' });

        expect(searchAppInputWrapper.element).toBeTruthy();
        expect(searchAppInputWrapper.attributes('value')).toEqual(searchTerm);
    });

    it('Should bind "selectedSearchByOption" getter to search by options', () => {
        //Arrange
        const selectedSearchByOption: SearchByOptionNames = SearchByOptionNames.Genre;

        const getters = {
            [getFilmModuleType(FilmGetterProps.SearchTerm)]: (): string => '',
            [getFilmModuleType(FilmGetterProps.SearchBy)]: (): SearchByOptionNames => selectedSearchByOption,
        };

        const store = new Vuex.Store({
            getters,
        });

        const wrapper = shallowMount(SearchPanel, {
            store,
            localVue,
        });

        //Act & Assert
        const searchAppInputWrapper = wrapper.findComponent({ name: 'button-group' });

        expect(searchAppInputWrapper.element).toBeTruthy();
        expect(searchAppInputWrapper.attributes('selectedoption')).toEqual(selectedSearchByOption);
    });
});
