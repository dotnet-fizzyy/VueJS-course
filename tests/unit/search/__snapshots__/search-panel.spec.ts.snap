// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPanel.vue Tests Should render 1`] = `
<div
  class="root"
  data-aqa-search-panel=""
>
  <span
    class="main-label"
  >
    Find your movie
  </span>
   
  <div
    class="search-container"
  >
    <div
      class="input-container"
    >
      <app-input-stub
        fontsize="24"
        placeholder="Search"
        value="value"
      />
    </div>
     
    <div
      class="search-button-container"
    >
      <primary-button-stub
        fontsize="24"
        fontstyle="normal"
      >
        Search
      </primary-button-stub>
    </div>
  </div>
   
  <div
    class="options-container"
  >
    <button-group-stub
      label="Search By"
      options="[object Object],[object Object]"
      selectedoption="g_e"
    />
  </div>
</div>
`;
